### DISCUSSION TITLE
Tentative list of deprecations and breaking changes for Astro 3.0

### DISCUSSION THREAD

**Discussion by Princesseuh** [withastro/roadmap#502] [Proposal]
### Body

As we're heading towards 3.0, @matthewp and I thought it would be neat to have a discussion / document talking about the plans regarding deprecations in Astro 3.0. Previously, we would've had this in private among us maintainers, but we figured there was no downsides to having this public. So here we are.

> **Warning** **for users**
> This is a tentative list, some things will happens, some won't, some will that are not documented here. Do not change your code or make any plans based on the things in this document.

# Core

- [ ] Investigate dropping support for returning simple objects from endpoints. Returning `Response`s is better in all cases apart from one (SSG, files that need specific encoding) and supporting both is leading to user confusion, maintenance burden and bugs.
- [ ] Move `astro check` to a separate package, or at least its TypeScript dependency, in order to significantly reduce the size of the `astro` package and allow users more autonomy over the TypeScript version.

**Comment by matthewp** [comment]
Something to think / talk about. CSS ordering bugs continue to happen. We need to investigate if there's an architectural issue where we cannot ensure the exact same CSS order between dev and build. This would be because of code splitting.

If it turns out we cannot fix this, then we need to reconsider where to use `:where()` or not. Using class names would prevent most of these kinds of bugs, even though CSS order might not match dev.

**Reply by ematipico** [reply]
CSS ordering has always been a pain, even in `webapck` land. I am not very familiar yet with `astro` architecture.

Sometimes one of the issues is caused by the order when importing components. Given the following example, and assuming that the components use `import "style.css"` in their code:
```js
// page1.js
import ComponentA from "componentA.js"
import ComponentB from "componentB.js"
```

```js
// page2.js
import ComponentB from "componentB.js"
import ComponentA from "componentA.js"
```

This might result in a problem where we need to know which styles come first.

**Reply by matthewp** [reply]
Yes, this is exactly the sort of problem we need to see if we can solve. If we cannot then it would make sense to change back to using class names.

**Reply by jasikpark** [reply]
Asocciated Vite issue for a similar problem: https://github.com/vitejs/vite/issues/3924#issuecomment-1433031478 - this also affects me on a Vite + React project using CSS Modules: enabling code splitting triggers the ordering problem.

Though I may be discussing a different problem, CSS + JS modules are a messy task.

**Reply by matthewp** [reply]
Started a discussion on ways to fix this: https://github.com/withastro/roadmap/discussions/510

**Comment by matthewp** [comment]
# Implicit head injection

Astro will inject our head content (links and scripts) into the head, even if your page never has an explicit head tag. This causes bugs, especially with the introduction of content collections and its head propagation. Knowing that all pages have an explicit `<head>` tag would completely prevent all of these bugs. 

We could create a `Astro.injectHead()` method or something, for people who want manual control over injecting Astro's head content.

**Comment by ematipico** [comment]
> Move astro check to a separate package, or at least its TypeScript dependency, in order to significantly reduce the size of the astro package and allow users more autonomy over the TypeScript version.

What's the reason for removing the TS dependency?

**Reply by Princesseuh** [reply]
The idea would be to make it an optional peer deep. The idea is to make `astro` (the package) smaller by avoiding to install the language-server and TypeScript (70mb~) for everybody despite not everyone using `astro check`

**Reply by ematipico** [reply]
I see. I come from a different background, where the language server is just a consumer of the API that checks for diagnostics. And - ideally - this API lives inside a "core" package, where the core is `astro`. But the use cases here might be different.

**Reply by Princesseuh** [reply]
Ah yes, this is different here because the "API that checks for diagnostics" is TypeScript, which Astro itself does not depend on. For the actual Astro-related diagnostics, the language-server does depend on a core package, the compiler.

Having a non-dev dependency on TypeScript is generally unwanted, because you want users to be able to control the version. The version of it might even be provided by their editor, for instance

**Comment by natemoo-re** [comment]
A feature we had in the Snowpack days but dropped in the switch to Vite was requiring the `node:` prefix to access Node built-ins in user code. If we could bring that back, I think it would be a great change. 

Reason: I've spent a bunch of time debugging very weird issues with folks only to realize that they had installed a package that used the `fs` (or similar) specifier, overwriting the built-in `node:fs` package. Enforcing the prefix would eliminate this class of mistakes and future-proof us for non-Node runtimes.

**Comment by matthewp** [comment]
Endpoints have a legacy 2 argument form that we still have code for in the codebase. We can remove that in 3.0

**Comment by matthewp** [comment]
3.0 is out...
